python/ray/data/preprocessor.py

```diff
@@ -234,42 +240,54 @@ def _transform(self, dataset: Dataset) -> Dataset:
             return dataset.map_batches(self._transform_pandas, batch_format="pandas")
         elif transform_type == "arrow":
             return dataset.map_batches(self._transform_arrow, batch_format="pyarrow")
+        elif transform_type == "numpy":
+            return dataset.map_batches(self._transform_numpy, batch_format="numpy")
         else:
             raise ValueError(
                 "Invalid transform type returned from _determine_transform_to_use; "
-                f'"pandas" and "arrow" allowed, but got: {transform_type}'
+                f'"pandas", "arrow" and "numpy" allowed, but got: {transform_type}'
             )
 
-    def _transform_batch(self, df: "DataBatchType") -> "DataBatchType":
+    def _transform_batch(self, data: "DataBatchType") -> "DataBatchType":
         import pandas as pd
+        import numpy as np
 
         try:
             import pyarrow
         except ImportError:
             pyarrow = None
 
-        if isinstance(df, pd.DataFrame):
+        if isinstance(data, pd.DataFrame):
             data_format = "pandas"
-        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+        elif pyarrow is not None and isinstance(data, pyarrow.Table):
             data_format = "arrow"
+        elif np is not None and isinstance(data, (dict, np.ndarray)):
+            data_format = "numpy"
         else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames "
-                f"and PyArrow Tables. Got {type(df)}."
+                "`transform_batch` is currently only implemented for Pandas "
+                f"DataFrames, PyArrow Tables and Numpy ndarray Got {type(data)}."
             )
 
         transform_type = self._determine_transform_to_use(data_format)
 
         if transform_type == "pandas":
             if data_format == "pandas":
-                return self._transform_pandas(df)
+                return self._transform_pandas(data)
             else:
-                return self._transform_pandas(df.to_pandas())
+                return self._transform_pandas(data.to_pandas())
         elif transform_type == "arrow":
             if data_format == "arrow":
-                return self._transform_arrow(df)
+                return self._transform_arrow(data)
+            else:
+                return self._transform_arrow(pyarrow.Table.from_pandas(data))
+        elif transform_type == "numpy":
+            if data_format == "numpy":
+                return self._transform_numpy(data)
             else:
-                return self._transform_arrow(pyarrow.Table.from_pandas(df))
+                raise NotImplementedError(
+                    "Not sure how to handle non-numpy data foramt yet."
+                )
```

Comment:
I think we'd want to do the same conversion semantics as our data batch conversion, i.e. a single-column table is turned into a single ndarray, while a multi-column table is turned into a dict of ndarrays: https://github.com/ray-project/ray/blob/ef260702a20c8424ebc7914bf72800e5c409fb29/python/ray/air/util/data_batch_conversion.py#L86-L95